Motion Matching as an AnimationRootNode

GuilhermeGSousa opened the pull request:
- [x] Turn `MMQueryInput` into a `RefCounted` object
- [x] Add `MMQueryInput` as a animation node parameter
- [x] Run a motion matching query on the motion matching node's animation library
- [x] Figure out how `blend_animation` works so that we can play/blend the animation result
- [x] Play/Blend the animation result

fire commented:
What do you want me to test? Is it complete?

> Figure out how blend_animation works so that we can play/blend the animation result
>
> Play/Blend the animation result

These are not part of the AnimationRootNode but blending systems of the AnimationTree

fire commented:
@TokageItLab Can you take a look?

GuilhermeGSousa commented on src/mm_character.cpp:
```diff
@@ -329,26 +331,33 @@ void MMCharacter::_update_query(double delta_t) {
 
-        // Run query
-        const MMQueryOutput result = query(query_input);
+        // // Run query
```
@fire the part that is missing is to bring this code over to `MMAnimationNode`. The thing is, the animation node can't just do `_animation_player->play(...)`, it needs to call `AnimationNode::blend_animation`, much like what `AnimationNodeAnimation` does. This is the API I'm not as familiar with, I'll try to play around with it a bit today.

GuilhermeGSousa commented:
This last commit show do a better job of showing what I'm trying to do here

fire commented:
Do you have a test project? Without blending?

GuilhermeGSousa commented:
@fire pushed my demo changes to https://github.com/GuilhermeGSousa/godot-motion-matching-demo/tree/mm-animation-node

fire commented:
Is https://github.com/GuilhermeGSousa/godot-motion-matching/pull/42/commits/d123855cef907c40a4be7c49b8dd0df54e9ec57d usable yet?

It says first functioning node.

GuilhermeGSousa commented:
@fire everything you said is correct, it is useable with its corresponding branch on the demo repo. And yes, we need to find a way to implement blending, I do have some ideas but I'm no expert in animation tree code, still learning how all this works

GuilhermeGSousa commented:
Cleaned things up a bit more. On the demo side, I removed the DampedSkeletonModifier, since that "hides" the fact we're not actually blending animations together on the AnimationTree
